modules/ocl/cv_feature_match_cluster.cpp

```diff
@@ -0,0 +1,351 @@
+/*
+ * cv_feature_match_cluster.cpp - optical flow feature match selected by clustering
+ *
+ *  Copyright (c) 2016-2017 Intel Corporation
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *   http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ *
+ * Author: Wu Junkai <junkai.wu@intel.com>
+ */
+
+#include "cv_feature_match_cluster.h"
+#include "xcam_obj_debug.h"
+#include "image_file_handle.h"
+#include "cl_utils.h"
+
+#define XCAM_CV_FM_DEBUG 0
+#define XCAM_CV_OF_DRAW_SCALE 2
+
+namespace XCam {
+CVFeatureMatchCluster::CVFeatureMatchCluster ()
+    : CVFeatureMatch ()
+{
+    XCAM_ASSERT (_cv_context.ptr ());
+}
+
+bool
+CVFeatureMatchCluster::calc_mean_offset (
+    std::vector<cv::Point2f> &corner0, std::vector<cv::Point2f> &corner1,
+    std::vector<uchar> &status, std::vector<float> &error,
+    float &mean_offset_x, float &mean_offset_y,
+    cv::InputOutputArray debug_img, cv::Size &img0_size, cv::Size &img1_size)
+{
+    std::vector<std::vector<uint32_t>> clusters;
+    std::vector<std::vector<cv::Point2f>> clusters_offsets;
+    std::vector<uint32_t> valid_seeds (status.size ());
+    std::vector<uint32_t> valid_corners (status.size ());
+
+    for (uint32_t i = 0; i < status.size (); ++i) {
+        if (!status[i] || (error[i] > _config.max_track_error) || corner1[i].x < 0.0f || corner1[i].x > img0_size.width) {
+            valid_corners[i] = 0;
+            valid_seeds[i] = 0;
+        } else {
+            valid_corners[i] = 1;
+            valid_seeds[i] = 1;
+        }
+    }
+
+    float seed_x_offset = 0.0f;
+    float seed_y_offset = 0.0f;
+
+    std::vector<uint32_t> cluster;
+    std::vector<cv::Point2f> cluster_offset;
+
+    float thres = 8.0f;
+    while (1) {
+        for (uint32_t i = 0; i < status.size (); ++i) {
+            if (valid_seeds[i]) {
+                seed_x_offset = corner1[i].x - corner0[i].x;
+                seed_y_offset = corner1[i].y - corner0[i].y;
+                cluster.push_back (i);
+                cluster_offset.push_back (cv::Point2f(seed_x_offset, seed_y_offset));
+                valid_corners[i] = 0;
+                valid_seeds[i] = 0;
+                break;
+            }
+        }
```

Comment:
is it possible to cause a dead loop here while(1). and something like kwscan may report issue either.

Reply:
It won't cause a dead loop but I'll modify the condition to be explicit anyway.